### ISSUE TITLE
7.0.1 patch release

### ISSUE THREAD

**Issue by peastman** [pandegroup/openmm#1489]
I think we ought to put out a 7.0.1 release to fix a few bugs.  #1487 definitely should be included.  Probably also #1484 and #1466.  There have been a few other bug fixes since 7.0, but those are the only ones that I think are really important, and keeping the changes to a minimum means we can get it out faster with less risk of breaking something in the process.

**Comment by mpharrigan** [comment]
I've pushed `7.0_branch` here https://github.com/pandegroup/openmm/tree/7.0_branch

It includes #1484 and #1466. I can cherry-pick #1487 once it's merged

**Comment by peastman** [comment]
Thanks!  #1487 is now merged.

**Comment by mpharrigan** [comment]
Cherry-picked that as well.

I'd like to get more involved with the release process. As I understand it, Jenkins is used to build Source, Linux, and Mac distributions. Travis-CI is used to make Linux and Mac conda distributions. Peter makes the windows releases on a local machine. Is this correct?

Should we make the source distribution first and use that for the conda recipe, rather than relying on git tags or hashes?

**Comment by jchodera** [comment]
> As I understand it, Jenkins is used to build Source, Linux, and Mac distributions. 

Currently, yes, but I want to migrate away from Jenkins for these.  I've proposed before that we add travis branches to build these packages every commit to `master` and upload the artifacts to S3---I still think this is the right idea!

> Travis-CI is used to make Linux and Mac conda distributions. 

Yes, through the http://github.com/omnia-md/conda-recipes repo

> Peter makes the windows releases on a local machine. Is this correct?

Yep. There's no way to get OpenCL / CUDA onto AppVeyor right now, and we haven't played with other cloud services that might permit this.

**Comment by peastman** [comment]
Thanks!  At this point I think we're all ready to do the builds?

> Is this correct?

I think so?  @jchodera can confirm the details.

> Should we make the source distribution first and use that for the conda recipe, rather than relying on git tags or hashes?

We've discussed that before, though I don't remember if we reached any resolution on the question.  That's how I used to do it back when I was doing all the builds by hand on VMs.  Since this is a patch, though, we should try to do the build as close as possible to exactly the same way we did 7.0.  The less we change, the fewer chances there are to introduce new problems, and patch releases do go through the extended testing process that full releases do.

**Comment by peastman** [comment]
> patch releases do go through

That should have been, "don't go through."

**Comment by jchodera** [comment]
Note that we are now **following this release process strictly**:

https://github.com/pandegroup/openmm/wiki/Release-Candidates

Now is the time to update this if we want to deviate from this at all.

**Comment by jchodera** [comment]
I would propose one modification: We call the version `7.0.1` not `7.0.1rcZ`.

**Comment by mpharrigan** [comment]
Ok. For 7.1, we can discuss switching to building a source version first, and binaries from that. Now, everything is built from a git revision. Specifically, we use a git hash that only gets tagged after a release candidate is publicly available. (correct me if I'm wrong)

**Comment by jchodera** [comment]
> Now, everything is built from a git revision. Specifically, we use a git hash that only gets tagged after a release candidate is publicly available. (correct me if I'm wrong)

Correct, according to [the wiki](https://github.com/pandegroup/openmm/wiki/Release-Candidates)

**Comment by peastman** [comment]
Yes, the wiki description is correct.  But unlike a major release where we go through a beta process first, with patches we jump straight to the release candidate.  The main point of betas is to let users try out the new features, and since a patch doesn't have any new features, that's unnecessary.

**Comment by peastman** [comment]
> I would propose one modification: We call the version 7.0.1 not 7.0.1rcZ.

Are you referring to the git tag?  If so, what do we do if we then decide we need to do a second release candidate?

If the answer is, "We move the git tag so it now points to a different revision," I'm ok with that.  But you've objected to that in the past.

**Comment by mpharrigan** [comment]
I think he's referring to the conda package version

**Comment by jchodera** [comment]
> Yes, the wiki description is correct. But unlike a major release where we go through a beta process first, with patches we jump straight to the release candidate

So what you're really saying is "the wiki description is incorrect".  :)

Can you amend the wiki description so that it is correct?

> Are you referring to the git tag? 

No, the conda version.

**Comment by peastman** [comment]
No, the wiki page you linked to is titled "release candidates", and it is a completely accurate description of how to build release candidates.  It never mentions anything about betas.  What part of it do you think is incorrect?

**Comment by jchodera** [comment]
Now I'm more confused.  The page mentions "Typically that involves a post to the discussion forum and an email to the beta testers mailing list." We have beta testers test release candidates?  Or no?  Are betas different from release candidates?  Is there a separate page on betas?  Or when we build betas vs release candidates, or proceed from betas to release candidates?

To make this simpler, just to triple-confirm: We're following the instructions [on this page](https://github.com/pandegroup/openmm/wiki/Release-Candidates) to the letter, right?  If so, great---we just need to know who is doing what.

**Comment by peastman** [comment]
> We have beta testers test release candidates?

Of course.  The mailing list (openmm-betatesters@simtk.org) is for people who have volunteered to test prerelease software.  Just because the list name has "betatesters" in it, that doesn't mean the people on it aren't interested in release candidates.

> Are betas different from release candidates?

Very.  See https://github.com/pandegroup/openmm/wiki/Cutting-an-OpenMM-release for an overview of the extended release process (not very detailed, could use a lot more information, but it's what's there right not).

> We're following the instructions on this page to the letter, right?

Unless you see something there that you think should be changed, that's the plan!

**Comment by jchodera** [comment]
Nothing to change! I just can't be very helpful in this process unless everything is spelled out in pedantic detail due to other commitments.

**Comment by peastman** [comment]
Ok!  Are we all ready to start building?  It should be the latest revision in the 7.0 branch, which is 5e86c4f76cb8e40e026cc78cdc452cc378151705.

**Comment by jchodera** [comment]
Again, sorry to be pedantic here, but I believe we needed a clear assignment of tasks to people since the procedure doesn't yet include that information and @mpharrigan wants to be more involved this time.

**Comment by peastman** [comment]
I'll do the Windows packages, you do the Linux, Mac, and source packages.  Since this is a patch, we should do everything as close as possible to exactly the same way we did 7.0.  Any change, like having a certain task done by a different person on a different computer, is an opportunity for something new to go wrong.

**Comment by jchodera** [comment]
OK.

What git hash are we building to?

**Comment by peastman** [comment]
See https://github.com/pandegroup/openmm/issues/1489#issuecomment-220168110. :)

**Comment by jchodera** [comment]
Sorry. I keep asking everything be laid out specifically because I am juggling so many things and information not readily at hand makes it difficult for me to pitch in. This was obviously a dumb oversight on my part, but just wanted you to know why I keep asking for more organization and checklists.

**Comment by peastman** [comment]
Here's a checklist for marking things off as they get done.

- [ ] Build Linux zip installer (John)
- [ ] Build Mac zip installer (John)
- [ ] Build Windows zip installer (Peter)
- [ ] Build Linux conda installer (John)
- [ ] Build Mac conda installer (John)
- [ ] Build Windows conda installer (Peter)
- [ ] Build source package (John)
- [ ] Test Linux zip installer (anyone)
- [ ] Test Mac zip installer (anyone)
- [ ] Test Windows zip installer (anyone)
- [ ] Test Linux conda installer (anyone)
- [ ] Test Mac conda installer (anyone)
- [ ] Test Windows conda installer (anyone)

**Comment by leeping** [comment]
Hi there,

I just wanted to add that the latest commits fixed an error that we saw when running an OpenMM simulation using AMBER inputs which had a triclinic box and a trunctated octahedron of water molecules.  The kinetic temperature was 1000K previously even though we had a set temperature of 300K.  Now the kinetic temperature is 300K as expected.

Thanks, 

- Lee-Ping

**Comment by jchodera** [comment]
Thanks for the checklist! Just getting back to NYC now, but all of our build machines seem to have gone offline in the week I've been away. Unfortunately this means I can't build anything until Monday unless someone happens to head into the lab over the weekend.

**Comment by peastman** [comment]
Any progress on this?

**Comment by jchodera** [comment]
Hardware is back up, but MSK is making us upgrade Jenkins because of site vulnerabilities. Spent the last day trying to do the upgrade. No functioning Jenkins yet.

Sorry about this. We really need to eliminate the Jenkins part of the build system.

**Comment by peastman** [comment]
Sigh.  Yeah, we need to figure out a better way of doing this.

**Comment by jchodera** [comment]
Should have this up and working again today.

**Comment by jchodera** [comment]
* [`osx` build](https://jenkins.choderalab.org/job/package-openmm-osx/ws/packaging/)
* `linux` build (pending)
* source build (pending)

**Comment by jchodera** [comment]
We may have run into a problem with the conda installers due to our use of `7.0rc1`:
https://github.com/omnia-md/conda-recipes/pull/531

**Comment by jchodera** [comment]
@peastman: Looks like the [linux](https://jenkins.choderalab.org/job/package-openmm-linux-vagrant/ws/packaging/) and [source](https://jenkins.choderalab.org/job/package-openmm-source/ws/packaging/) zip installers ended up with `OpenMM-7.0.0` names instead of 7.0.1.  They both use [this logic](https://github.com/pandegroup/openmm/blob/master/devtools/packaging/scripts/linux/package.sh#L8) to grab the version from `build/CMakeCache.txt`.  Did you remember to update the OpenMM version number?

**Comment by jchodera** [comment]
Whoops, the issue instead seems to be with [not finding the right git hash](https://jenkins.choderalab.org/job/package-openmm-source/74/consoleFull):
```
+ vagrant ssh --command 'rm -rf openmm && git clone -b 5e86c4f https://github.com/pandegroup/openmm.git'
Initialized empty Git repository in /home/vagrant/openmm/.git/
warning: Remote branch 5e86c4f not found in upstream origin, using HEAD instead
```

**Comment by jchodera** [comment]
Rebuilding. I am worried this may have been a problem for 7.0rc1 as well, but I don't know which Jenkins build number produced the artifact you are actually distributing.

**Comment by jchodera** [comment]
I have switched to
```
vagrant ssh --command 'rm -rf openmm && git clone -o 5e86c4f --depth=1 https://github.com/pandegroup/openmm.git'
```

**Comment by peastman** [comment]
Do you know why it isn't finding the hash?  Is that because the hash doesn't exist in the main branch, only in the 7.0 branch?  If so, that wasn't a problem for 7.0rc1 since we built that from the main branch.

**Comment by peastman** [comment]
Perhaps let's stick to only one checklist?  The one I posted above already has items for all the builds.  Creating a second checklist with some of the same items will confuse all of us.

**Comment by peastman** [comment]
The Jenkins build for Mac is labeled 7.0.1 as expected, but the Linux and source ones are labeled 7.0.0.  Are you sure they're really the right versions?

**Comment by peastman** [comment]
By the way, I just made a small change to https://github.com/pandegroup/openmm/wiki/Release-Candidates.  I specified that when we test builds, we should check the values of `Platform.getOpenMMVersion()` and `simtk.openmm.version.git_revision`.  That seems like a useful sanity check.

**Comment by jchodera** [comment]
> Perhaps let's stick to only one checklist? The one I posted above already has items for all the builds. 

As I said, I can't edit your checklist.

**Comment by jchodera** [comment]
> Do you know why it isn't finding the hash?

I imagine this was because the ancient version of `git` in the `holy-build-box` didn't support using short hashes in the `-b` argument, whereas the `-o` argument works just fine.

**Comment by jchodera** [comment]
> The Jenkins build for Mac is labeled 7.0.1 as expected, but the Linux and source ones are labeled 7.0.0. Are you sure they're really the right versions?

Maybe the `-o` argument isn't working right either. Sigh.

**Comment by jchodera** [comment]
Rebuilding using a different syntax.

**Comment by jchodera** [comment]
@peastman: As noted above, I believe the source and linux ZIP installers for 7.0.0 correspond to the wrong git hash, since the ancient `git` behavior seems to have been to ignore the specified git hash and pull from `head` instead. When I checked the last working builds from 30 Mar 2016, I see
```
+ vagrant ssh --command 'rm -rf openmm && git clone -b f433936 https://github.com/pandegroup/openmm.git'
Initialized empty Git repository in /home/vagrant/openmm/.git/
warning: Remote branch f433936 not found in upstream origin, using HEAD instead
```

**Comment by mpharrigan** [comment]
why not do the sensible thing and `git clone && cd openmm && git checkout <hash>`

**Comment by mpharrigan** [comment]
The manpages says `-b` is for branches and `-o` just renames the default branch from `origin` to the name passed

**Comment by mpharrigan** [comment]
And shouldn't we use the long hashes to be safe?

**Comment by peastman** [comment]
I just checked OpenMM-7.0.0-Linux.zip, and it looks ok.  Its version.py file specified

```Python
git_revision = 'f4339363dcc4c904b2baae05949dedc0676b3f50'
```

The matches the version specified at https://github.com/pandegroup/openmm/releases for 7.0.

**Comment by peastman** [comment]
> As I said, I can't edit your checklist.

Then I'll mark things off as people complete them.  The whole point of having a checklist is for everything to be in one place.

**Comment by jchodera** [comment]
> Then I'll mark things off as people complete them. The whole point of having a checklist is for everything to be in one place

I've converted these to bullet points.

**Comment by jchodera** [comment]
>  I specified that when we test builds, we should check the values of `Platform.getOpenMMVersion()` and `simtk.openmm.version.git_revision`. 

Should we build these checks into the conda packaging tests as well?
```bash
python -c "import simtk ; print simtk.openmm.version.git_revision"
python -c "from simtk import openmm ; print openmm.version.git_revision"
```

**Comment by jchodera** [comment]
Here is the [updated linux ZIP installer](https://jenkins.choderalab.org/job/package-openmm-linux-vagrant/ws/packaging/)

**Comment by jchodera** [comment]
Here is the [updated source ZIP installer](https://jenkins.choderalab.org/job/package-openmm-source/ws/packaging/)

**Comment by jchodera** [comment]
I've added the tests to the conda packaging and merged the PR, so the new conda packages should appear on the `rc` channel soon.

**Comment by peastman** [comment]
Thanks!  The zip installers are now posted at https://simtk.org/frs/?group_id=161.  (Click "Hidden Releases" to see 7.0.1.)  Now we need to test them.  @mpharrigan do you want to help with that?  The things to check are

- The installation should run successfully.
- `python -m simtk.testInstallation` should run without errors, and all platforms should be present.
- `Platform.getOpenMMVersion()` and `simtk.openmm.version.git_revision` should have the correct values.

**Comment by peastman** [comment]
Why did you add `skip: True # [win]` in meta.yaml?  That's preventing me from building the Windows conda installers.

**Comment by jchodera** [comment]
Because the logic that *had* been in there to try to skip on `conda`:
```YAML
skip:
   # don't build on Appveyor (no OpenCL / CUDA)
   - [ {{ os.environ.get('APPVEYOR', 'False') == 'True' }} ]
```
was invalid syntax:
```
WARNING openmm Error: Invalid selector in meta.yaml line 13:
    - [ {{ os.environ.get('APPVEYOR', 'False') == 'True' }} ]
```
and I found nothing in the [conda preprocessor documentation](http://conda.pydata.org/docs/building/meta-yaml.html#preprocessing-selectors) that had anything close to this kind of syntax. I don't know where it came from, and I don't know if it *ever* worked.

I figured that since you were building the `win` conda package by hand anyway, the five seconds required to manually fix it was a small price to pay for not having to manually build the `osx` and `linux` packages as well.

**Comment by jchodera** [comment]
If you do figure out the selector syntax (or could point me to some relevant documentation), I'm happy to put it back in and fix it up!

**Comment by peastman** [comment]
A heads-up that you expected me to modify the file by hand would have been appreciated!

I'm now getting an error on `print simtk.openmm.version.git_revision`.  That line assumes Python 2 syntax.

**Comment by jchodera** [comment]
> A heads-up that you expected me to modify the file by hand would have been appreciated!

I haven't finished the recipe yet.  Will let you know when it is not broken.

**Comment by jchodera** [comment]
What should the two additional tests report?  I get:
```
+ python -c 'from simtk import openmm ; print openmm.version.git_revision'
5e86c4f76cb8e40e026cc78cdc452cc378151705
+ python -c 'from simtk import openmm; print openmm.Platform.getOpenMMVersion()'
7.0
```

**Comment by peastman** [comment]
That's the right revision.  `Platform.getOpenMMVersion()` only reports (major).(minor).  Possibly we should modify it to give the patch version as well if that's non-zero, but that can wait for a future release.

**Comment by jchodera** [comment]
OK.  Still testing, hang tight.

**Comment by jchodera** [comment]
Well, the `openmm` recipe works now, but the package went to `main` instead of the `rc` label so I had to delete it immediately.  See https://github.com/omnia-md/conda-recipes/issues/535

You should be able to use the current [`openmm` recipe](https://github.com/omnia-md/conda-recipes/tree/master/openmm) to build the `win` package manually, @peastman, with the caveat that you still need to manually remove the `skip` section.

**Comment by jchodera** [comment]
The conda packages are finally ending up in the right place:
https://anaconda.org/omnia/openmm/files

**Comment by jchodera** [comment]
Looks like the `osx` and `linux` conda packages are all built now.

**Comment by peastman** [comment]
Thanks!  Building the Windows ones right now.

Now we need to test them.  Who wants to test which packages?

**Comment by peastman** [comment]
The Windows conda packages are uploaded, but for some reason it sent them to the label `omnia` instead of `rc`.  I don't know why.  And I don't seem to have permission to fix them.  @jchodera could you do that?

**Comment by jchodera** [comment]
Fixed.

**Comment by jchodera** [comment]
I've been trying to figure out how to give permission to other people, but I don't know if the current new UI supports that.

**Comment by peastman** [comment]
I tested the Windows build (both zip and conda).  Everything looks ok.

**Comment by peastman** [comment]
I verified the Linux installers.  That just leaves the Mac ones.  I need someone else to test those, because I don't have a Mac that supports CUDA.

**Comment by jchodera** [comment]
I think we only have the machine that built the Mac installer as an `osx` platform that supports CUDA, so we should probably test it on something else.

**Comment by peastman** [comment]
I tested on my Mac and verified that everything other than CUDA works.  Maybe go ahead and try it on the one you used for building?  If that's all we've got, at least it's better than nothing.

**Comment by peastman** [comment]
Does anyone have a Mac that supports CUDA they can test on?  This is now the only thing holding up the release.

**Comment by andrrizzi** [comment]
Unfortunately, we (choderalab) don't have any Mac with NVIDIA cards. The ones we have that support CUDA are all linux machines.

**Comment by peastman** [comment]
What about the one you built it on?

**Comment by jchodera** [comment]
Sorry about that. @andrizzi pointed out that it has CUDA installed but no CUDA-capable NVIDIA hardware. I had thought it did.

**Comment by jlmaccal** [comment]
I can try this. What do you want me to do?

**Comment by jchodera** [comment]
Thanks!

Instructions:

ZIP installer: https://jenkins.choderalab.org/job/package-openmm-osx/ws/packaging/
Testing protocol (see Step 3): https://github.com/pandegroup/openmm/wiki/Release-Candidates
OpenMM issue to report in: https://github.com/pandegroup/openmm/issues/1489#issuecomment-222005444

Thanks!

**Comment by peastman** [comment]
Thanks! For each Mac package (zip and conda), just install it, then run
testInstallation and make sure the CUDA platform works.

**Comment by jlmaccal** [comment]
OK, tests pass for both zip and conda 7.0.1 packages!

I'm not sure if all of the values in `mm.version` are correct though. For example, it has `release=False`.

**Comment by peastman** [comment]
Thanks!  I'll go post the announcement now.

**Comment by peastman** [comment]
I'm not sure how much longer we want to wait before officially releasing it.  As far as I can tell from the anaconda status page, it's not clear anyone has actually installed it other than us testing it.  But getting it out as soon as possible would be a good thing.

**Comment by jchodera** [comment]
What precisely are you waiting on? Feedback from who?

**Comment by jchodera** [comment]
I wonder if it's possible for us to add both the 'rc' label and the normal 'main' label to our conda channels for Travis testing so we can at least automate the testing of release candidate packages with the nightly builds that are triggered. Will report back in a couple of hours if this is successful.

**Comment by peastman** [comment]
Anyone who chooses to test them! :)

**Comment by peastman** [comment]
Unless someone feels we should keep waiting, I say we go ahead and release it.  @jchodera if you agree with that, can you move the tags on the conda releases (or figure out how to make it so people other than you can do that)?  I'll then make the simtk.org packages visible and post the announcement.

**Comment by jchodera** [comment]
I've concluded that I *can* add both the `main` and `rc` channels to `.travis.yml` to test against whatever `rc` versions are available, and that `openmm-7.0.1` [works fine with `openmmtools`](https://travis-ci.org/choderalab/openmmtools/jobs/134318296). I'm hesitant to [merge that PR](https://github.com/choderalab/openmmtools/pull/106) because I'd simultaneously have to update all of our projects to use this scheme, and it still wouldn't let us test against development versions of upstream codes. There's probably a better way to automate testing of `openmm` release candidates against a battery of downstream packages using continuous integration and `omnia`, but we can figure that out later.

> Unless someone feels we should keep waiting, I say we go ahead and release it. @jchodera if you agree with that, can you move the tags on the conda releases (or figure out how to make it so people other than you can do that)? I'll then make the simtk.org packages visible and post the announcement.

Agreed!  I've just moved the packages from `rc` to `main`.

I looked this past weekend to see if there was a way to authorize others to do this, and I can't figure out how this is possible with the current UI. I can create access tokens for you to use the command-line API, but I don't know how you would [use the CLI](http://docs.anaconda.org/using.html#UsingLabelsInTheDevelopmentCycle) to do anything other than move *all* packages in `omnia` from `rc` to `main`, which is certainly not what we want.
